### PR TITLE
perf(agg): local accumulator with deferred scatter for BatchFill/BatchMerge

### PR DESCRIPTION
**Pull request by aunjgr**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #24097

## What this PR does / why we need it:
Replace row-at-a-time scatter writes in aggregation kernels with a two-phase approach: accumulate into a 255-slot L1-resident open-addressing hash table, then scatter-back once per distinct group. This eliminates random cache misses on the chunked state arrays during the hot loop.

Key design decisions:
- Local accumulator uses XOR-folded hash with linear probing (255 slots fits in L1)
- Deferred scatter-back with `lastX` caching minimizes `vector.MustFixedColAsSlice` calls
- Zero `unsafe` in `aggexec` package — all pointer manipulation confined to `vector.MustFixedColAsSlice` in the vector package
- `*[AggBatchSize]T` array pointer access for bounds-check elimination
- Specialized non-generic Decimal64/128 executors (`sum_decimal_fast.go`) avoid type switches and interface boxing
- Overflow checking conditional on type width (Decimal64→128 provably safe, Decimal128→128 checked when precision > 28)

Microbenchmark results (4096 rows, 64 groups):
- SUM(int64):       45.8µs → 5.4µs   (8.4x)
- SUM(float64):     45.4µs → 5.3µs   (8.6x)
- SUM(decimal64):   19.3µs → 6.2µs   (3.1x, specialized fast path)
- AVG(int64):       37.5µs → 6.7µs   (5.6x)
- AVG(decimal64):   52.0µs → 19.6µs  (2.7x)
- COUNT(column):    11.3µs → 4.1µs   (2.8x)
- COUNT/Merge:      750ns  → 110ns   (6.8x)
- MIN/MAX(int64):   comparable to COUNT path

### CODE REVIEW

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
## Multi-Angle Review

### Overall

The two-phase local-accumulator approach is well-designed — 255-slot open-addressing hash table fits comfortably in L1, probe termination is guaranteed by always having at least one empty slot, and the overflow fallback to direct scatter prevents degenerate behavior. Benchmark numbers (3-8x) are credible given the cache-miss elimination.

Three issues should be addressed before merge:

---

### 1. Missing `lastX` caching in sumavg2.go scatter-back loops

`count2.go` and `minmax2.go` correctly cache the chunk pointer in scatter-back:

```go
lastX := -1
var vals *[AggBatchSize]int64
for s := 0; s < nSlots; s++ {
    x := int(g >> aggBatchSizeShift)
    if x != lastX {
        lastX = x
        vals = chunkArr[int64](exec.state[x].vecs[0])
    }
    ...
}
```

But all 4 scatter-back loops in `sumavg2.go` (`batchFillSum`, `batchFillAvg`, and the decimal variants) call `chunkArr` unconditionally on every iteration:

```go
for s := 0; s < nSlots; s++ {
    sums := chunkArr[T](exec.state[x].vecs[0])  // redundant when x unchanged
    ...
}
```

Up to ~254 redundant `MustFixedColAsSlice` calls per scatter-back. Please add `lastX` caching for consistency and to avoid the unnecessary overhead.

---

### 2. `localAddSafe` detection is fragile

```go
var a A
switch any(a).(type) {
case types.Decimal64, types.Decimal128:
    exec.localAddSafe = true
}
```

This checks only the **argument type** `A`, not the `A→S` width relationship. The safety claim is:
- Decimal64→Decimal128: 255 × 10^18 < 10^38 ✓
- Decimal128→Decimal256: 255 × 10^38 < 10^76 ✓

This happens to be correct for current call sites (`[Decimal64, Decimal128]` and `[Decimal128, Decimal256]`), but a future instantiation of `[Decimal128, Decimal128]` would silently skip overflow checking.

Please add a comment at this site documenting the invariant (valid instantiations and required `sizeof(S) > sizeof(A)`), or restructure the check to verify the relationship directly.

---

### 3. Test coverage gaps

The `TestLocalAccumulatorOverflow` is good for the >255-group fallback path. Please consider adding:

- **Const vector test**: the `IsConst` branch in the local accumulator is untested — a single-value const vector with multiple groups would exercise it.
- **Decimal overflow edge case**: test `Decimal256→Decimal256` (the checked/`localAddSafe=false` path) with values near the precision limit to validate that overflow is correctly caught in the local buffer.

---

### Minor observations (non-blocking)

- `countStarExec.Merge` now delegates to `BatchMerge`, allocating a `[]uint64` per call. Not a hot path, but a slight regression for single-merge. Acceptable.
- The `chunkArr`/`PreExtend` invariant (all state vecs must have capacity ≥ AggBatchSize) is maintained in `GroupGrow` and `UnmarshalFromReader`, but has no runtime assertion. A debug-build check would reduce future regression risk.
- The local accumulator pattern is copy-pasted across 7 functions. Understood as necessary for Go generics constraints in perf-critical code, but worth a brief comment in each linking to the others so a future editor knows to update all copies.

**Review by ouyuanning — APPROVED** [review]

File: pkg/sql/colexec/aggexec/sum_decimal_fast.go [hunk 1]
```diff
@@ -303,62 +335,96 @@ func (exec *sumDecimal128FastExec) batchFill(offset int, groups []uint64, vector
...
 	vals := vector.MustFixedColNoTypeCheck[types.Decimal128](vec)
-	// constMask: 0 for const vectors (always index 0), -1 for flat (use full idx).
 	constMask := -min(1, len(vals)-1)
 	hasNull := vec.HasNull()
 	var np *bitmap.Bitmap
 	if hasNull {
 		np = vec.GetNulls().GetBitmap()
 	}
+	checked := exec.overflowCheck
+
```
[the remark this answers is not in the document]
Comment (aunjgr) [reply pkg/sql/colexec/aggexec/sum_decimal_fast.go:345]: This is a pre-existing design from PR #24098 (commit d0de2274f), not introduced by this PR. The threshold (`Width > 28`) was chosen deliberately: with max input value ≈ 10^28, overflow requires ~10 billion rows *in a single group* — practically unreachable in any real workload. The unchecked path provides measurable throughput gain for the common case (DECIMAL(18,x) financial columns).

**Comment by aunjgr** [comment]
All three items addressed in 4cce7b2df:

1. **lastX caching in sumavg2.go** — Added to all 4 scatter-back loops (`batchFillSum`, `batchFillAvg`, and both decimal variants). Now consistent with count2/minmax2.

2. **localAddSafe documentation** — Added comment documenting valid instantiations and the `sizeof(S) > sizeof(A)` invariant, with explicit warning about hypothetical `[Decimal128, Decimal128]`.

3. **Test coverage** — Added `TestConstVectorAccumulator` (const vector with SumInt64, MinInt64, SumDecimal64Fast) and `TestDecimal256Overflow` (verifies overflow is caught in the local buffer for the `localAddSafe=false` path).

**Review by XuPeng-SH — APPROVED** [review]
## Review: Approve ✓

High-quality performance PR. The L1-resident local accumulator with deferred scatter is well-suited to the problem — impressive 3-8x speedups across aggregation types. Edge cases (const vectors, nulls, overflow, >255 group fallback) are handled carefully, and the test coverage is solid.

### Minor suggestions

**1. Latent risk in `decimalStateAddUnchecked` for Decimal256**

`decimalStateAddUnchecked` silently discards the error from `Add256`. While `localAddSafe=false` for Decimal256→Decimal256 means this path is currently unreachable, if someone later changes the `localAddSafe` logic, overflow errors would be silently swallowed. Consider panicking in the Decimal256 branch or adding a comment documenting why it's dead code.

**2. `countStarExec.Merge` allocates per call**

```go
func (exec *countStarExec) Merge(...) error {
    return exec.BatchMerge(next, groupIdx2, []uint64{uint64(groupIdx1 + 1)})
}
```

This allocates a `[]uint64` on each `Merge` call. The old direct implementation was allocation-free. Consider restoring it if `Merge` is ever called in a hot loop.

**3. Test coverage gaps**

- No test for `batchFillAvg` with >255 groups (only SUM and COUNT overflow paths are tested).
- No test for `BatchMerge` with the new `chunkArr` paths.
- No test for null-heavy vectors (e.g., 50% nulls mixed with the overflow path).

**4. Maintainability note**

The local accumulator pattern is copy-pasted ~9 times (deliberate trade-off for performance over DRY — correct call). A brief comment listing all files that implement the pattern would help future maintainers apply changes consistently.
